Add binary representation to all bitvector values in the xml trace

thomasspriggs opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->
 Add binary representation to all bitvector values in the xml trace. The purpose of this is to provide a consistent interface for getting the values of bitvector-ed types out of the trace.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] N/A - None claimed - ~My commit message includes data points confirming performance improvements (if claimed).~
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

hannes-steffenhagen-diffblue commented on regression/cbmc/integral-trace/test.c:
```diff
@@ -0,0 +1,30 @@
+  assert(false);
+
+  return 0;
+}
```
missing newline

thomasspriggs replied:
Fixed.

hannes-steffenhagen-diffblue commented:
Looks like the new regression test is crashing with an invariant violation? Also, you should probably update [the actual XSD](https://github.com/diffblue/cbmc/blob/develop/doc/assets/xml_spec.xsd) with this.

thk123 commented on src/goto-programs/xml_goto_trace.cpp:
```diff
@@ -25,31 +25,32 @@ Author: Daniel Kroening
 xmlt full_lhs_value(const goto_trace_stept &step, const namespacet &ns)
 {
-  xmlt full_lhs_value{"full_lhs_value"};
+  xmlt xml{"full_lhs_value"};
```
:pick: Personally don't love this name either just cos the type tells you its xml /shrug

thomasspriggs replied:
Ok. I am going to make this `value_xml` instead.

martin-cs commented on doc/assets/xml_spec.tex:
```diff
@@ -164,7 +166,7 @@ \subsection{Trace Steps in XML}
   <type>signed int</type>
   <full_lhs>b</full_lhs>
-  <full_lhs_value>-1879048192</full_lhs_value>
+  <full_lhs_value binary="10010000000000000000000000000000">-1879048192</full_lhs_value>
```
This seems like a bit of a weird way of giving this.  As there are multiple possible representations of the value, wouldn't it be better as `<full_lhs_value_binary></>` `<full_lhs_value_decimal></>` etc.?  Or possibly with `full_lhs_value` be the default human-readable.  Then again, I am not an XML user so maybe this is the appropriate way of doing it.

thomasspriggs replied:
This is consistent with the way that the binary value is already given for floating point values. My intention is to add this feature in a way which will require fewer code changes for anyone who is processing the xml traces. If I was writing this representation from scratch, I would have probably chosen to given equal prominence to the binary and the decimal. This could looks something like -
```
<full_lhs_value><binary>10010000000000000000000000000000</binary><decimal>-1879048192</decimal></full_lhs_value>
```

martin-cs commented on regression/cbmc/integral-trace/test.desc:
```diff
@@ -0,0 +1,22 @@
+<full_lhs>u8</full_lhs>\n\s*<full_lhs_value binary="01100001">97</full_lhs_value>
+<full_lhs>u16</full_lhs>\n\s*<full_lhs_value binary="0{12}1000">8</full_lhs_value>
+<full_lhs>u32</full_lhs>\n\s*<full_lhs_value binary="0{16}0{11}10000">16u</full_lhs_value>
+<full_lhs>u64</full_lhs>\n\s*<full_lhs_value binary="0{32}0{16}0{10}100000">32ul</full_lhs_value>
```
FWIW : This is not what would expect to find in a field marked `binary`; I would just expect `0` and `1`.

hannes-steffenhagen-diffblue replied:
@martin-cs this: `0{32}` is regex notation for repetition, it means 32 0's. I assume this was chosen here to make the structure of the matched expressions somewhat easier to see. This isn’t actually in the output. Though I’m not sure what the rationale was of doing this over just `0{58}100000`

martin-cs replied:
Ah; yes.  I missed that it is a regex in the test pattern rather than a literal output.  Thanks.

martin-cs reviewed (approved):
Just thoughts; no action required.

thomasspriggs commented:
@hannes-steffenhagen-diffblue The xsd does not need to be updated, because it already contains the binary attribute. This was already present in the xsd because it is already present for floating point values.

thomasspriggs committed:
Add binary representation to all bitvector values in the xml trace

The purpose of this is to provide a consistent interface for getting the
values of bitvector-ed types out of the trace.

An existing test is updated in this commit in order for the existing
to test to match new functionality and so that every commit still passes
all tests.

There are instances when running with `--cprover-smt2` where the type of
the expression is `bitvector_typet`, but the expression is not a
`constant_exprt`. By removing the printing in this case we can avoid
breaking the partial functionality available with the `--cprover-smt2`
option.

thomasspriggs committed:
Add test of binary inclusion in xml trace for integer values

In order to confirm that the binary is now included in the xml trace
output.

Lines may be separated with `\r\n` on windows, rather than just `\n`. So
we need to match both kinds of separator in order to be platform
independant. As we don't require the xml elements to be on separate
lines and `\s` matches both `\r` and `\n` we can let `\s*` match all the
line separators as well as any indentation characters.

`uint32_t` for example fully specifies that an integer is unsigned and
32 bits. But it could be implemented as a `typedef unsigned` or a
`typedef unsigned long`, so long as these underlying types fulfill the
requirements. Therefore the regex for this needs to match both `16u`
and `16ul` as the value.

thomasspriggs committed:
Rename `full_lhs_value` to `value_xml`

In order to disambiguate it from `step.full_lhs_value`.

thomasspriggs committed:
Refactor to extract `step.full_lhs_value` into `value` variable

To avoid repeating ourselves and because "Which value?" is clear from
the context of the function.

thomasspriggs committed:
Explain why it is ok to use `false` for `is_signed`

To answer the question of why this is not a bug when reading the code.
Ideally there would exist a `bvrep2binary` function, but this is out of
scope of this PR.

thomasspriggs committed:
Move nil check to top of function

So that the filtering is done first and we avoid building temporaries
which aren't going to be used.

thomasspriggs committed:
Document the additional binary in the xml file

Because the PR checklist says that feature changes should be documented.

hannes-steffenhagen-diffblue commented on src/goto-programs/xml_goto_trace.cpp:
```diff
@@ -27,15 +27,15 @@ Author: Daniel Kroening
 xmlt full_lhs_value(const goto_trace_stept &step, const namespacet &ns)
 {
-  xmlt full_lhs_value{"full_lhs_value"};
+  xmlt value_xml{"full_lhs_value"};
```
I am not sure if I agree with the reasoning of this rename. I’d probably have gone for `full_lhs_value_xml` instead.

hannes-steffenhagen-diffblue commented on src/goto-programs/xml_goto_trace.cpp:
```diff
@@ -25,31 +25,34 @@ Author: Daniel Kroening
-  if(full_lhs_value_includes_binary(step, ns))
+  const auto &bv_type = type_try_dynamic_cast<bitvector_typet>(value.type());
+  const auto &constant = expr_try_dynamic_cast<constant_exprt>(value);
+  if(bv_type && constant)
```
I did like the `full_lhs_value_includes_binary` name, but I suppose if the intention is to use all bitvectors for this then this is more honest.

hannes-steffenhagen-diffblue reviewed (approved):
LGTM, thanks for addressing comments
